Variant call function is broken for multiple arguments

### **Description**
When calling the `call` function on a Variant and `args` has more than 1 element, the current running scene will crash.

`call` works correctly when `args` has only 1 element.

`call` function parameters for reference.
```
pub fn call(&mut self, method: &GodotString, args: &[Variant]) -> Result<(), ()>
```

### **Code snippet**
```
    #[export]
    // Publish to all listeners of the given event_key
    fn publish(&mut self, _owner: Node2D, event_key: GodotString, payload: Option<VariantArray>) {
        godot_print!("[INFO] Publishing");
        if !self.subscriptions.contains_key(&event_key.to_string()) {
            godot_error!("[ERROR] Key does not exist {:?}", event_key);
            return;
        }
        let listeners = self.subscriptions.get_vec(&event_key.to_string());
        let unwrapped_payload = payload.unwrap_or_else(|| VariantArray::new());
        match listeners {
            None => {
                godot_warn!("[WARN] Tried to publish to empty list");
            }
            _ => {
                let mut args = Vec::new();
                args.push(Variant::from_godot_string(&event_key));
                if !unwrapped_payload.is_empty() {
                    for item in unwrapped_payload.iter() {
                        args.push(item.clone());
                    }
                }
                godot_print!("args {:?}", args);
                for listener in listeners.unwrap().clone().iter_mut() {
                    // TODO currently broken
                    listener
                        .call(&GodotString::from_str("event_published"), args.as_slice())
                        .expect("[ERROR] Could not publish event");
                }
            }
        }
    }
```

--- a/gdnative-core/src/variant.rs
+++ b/gdnative-core/src/variant.rs
@@ -153,6 +153,36 @@ impl VariantType {
     }
 }
 
+#[repr(u32)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub enum CallError {
+    InvalidMethod =
+        sys::godot_variant_call_error_error_GODOT_CALL_ERROR_CALL_ERROR_INVALID_METHOD as u32,
+    InvalidArgument =
+        sys::godot_variant_call_error_error_GODOT_CALL_ERROR_CALL_ERROR_INVALID_ARGUMENT as u32,
+    TooManyArguments =
+        sys::godot_variant_call_error_error_GODOT_CALL_ERROR_CALL_ERROR_TOO_MANY_ARGUMENTS as u32,
+    TooFewArguments =
+        sys::godot_variant_call_error_error_GODOT_CALL_ERROR_CALL_ERROR_TOO_FEW_ARGUMENTS as u32,
+    InstanceIsNull =
+        sys::godot_variant_call_error_error_GODOT_CALL_ERROR_CALL_ERROR_INSTANCE_IS_NULL as u32,
+}
+
+impl CallError {
+    fn from_sys(v: sys::godot_variant_call_error_error) -> Result<(), CallError> {
+        if v == sys::godot_variant_call_error_error_GODOT_CALL_ERROR_CALL_OK {
+            Ok(())
+        } else {
+            debug_assert!(
+                (v as u32) <= sys::godot_variant_call_error_error_GODOT_CALL_ERROR_CALL_ERROR_INSTANCE_IS_NULL as u32,
+                "Godot should have passed a known error",
+            );
+
+            Err(unsafe { transmute(v as u32) })
+        }
+    }
+}
+
 // TODO: Looks like this is missing from the godot_headers bindings.
 // It's risky to redefine it here and count on the fact that the integer
 // constants will be the same.
@@ -490,32 +520,20 @@ impl Variant {
         unsafe { (get_api().godot_variant_has_method)(&self.0, &method.0) }
     }
 
-    // TODO: return a proper error.
-    pub fn call(&mut self, method: &GodotString, args: &[Variant]) -> Result<(), ()> {
+    pub fn call(&mut self, method: &GodotString, args: &[Variant]) -> Result<Variant, CallError> {
         unsafe {
             let api = get_api();
             let mut err = sys::godot_variant_call_error::default();
-            if args.is_empty() {
-                let mut first = ::std::ptr::null() as *const sys::godot_variant;
-                (api.godot_variant_call)(&mut self.0, &method.0, &mut first, 0, &mut err);
-            } else {
-                // TODO: double check that this is safe.
-                let gd_args: &[sys::godot_variant] = transmute(args);
-                let mut first = &gd_args[0] as *const sys::godot_variant;
-                (api.godot_variant_call)(
-                    &mut self.0,
-                    &method.0,
-                    &mut first,
-                    args.len() as i32,
-                    &mut err,
-                );
-            }
+            let mut arg_refs = args.iter().map(Variant::sys).collect::<Vec<_>>();
+            let variant = (api.godot_variant_call)(
+                &mut self.0,
+                &method.0,
+                arg_refs.as_mut_ptr(),
+                args.len() as i32,
+                &mut err,
+            );
 
-            if err.error == sys::godot_variant_call_error_error_GODOT_CALL_ERROR_CALL_OK {
-                Ok(())
-            } else {
-                Err(())
-            }
+            CallError::from_sys(err.error).map(|_| Variant::from_sys(variant))
         }
     }
 

--- a/test/src/lib.rs
+++ b/test/src/lib.rs
@@ -44,6 +44,8 @@ pub extern "C" fn run_tests(
     status &= test_rust_class_construction();
     status &= test_owner_free_ub();
 
+    status &= test_variant_call_args();
+
     gdnative::Variant::from_bool(status).forget()
 }
 
@@ -347,10 +349,101 @@ impl NativeClass for RegisterSignal {
 #[methods]
 impl RegisterSignal {}
 
+struct VariantCallArgs;
+
+impl NativeClass for VariantCallArgs {
+    type Base = Reference;
+    type UserData = user_data::MutexData<VariantCallArgs>;
+    fn class_name() -> &'static str {
+        "VariantCallArgs"
+    }
+    fn init(_owner: Reference) -> VariantCallArgs {
+        VariantCallArgs
+    }
+    fn register_properties(_builder: &init::ClassBuilder<Self>) {}
+}
+
+#[methods]
+impl VariantCallArgs {
+    #[export]
+    fn zero(&mut self, _owner: Reference) -> i32 {
+        42
+    }
+
+    #[export]
+    fn one(&mut self, _owner: Reference, a: i32) -> i32 {
+        a * 42
+    }
+
+    #[export]
+    fn two(&mut self, _owner: Reference, a: i32, b: i32) -> i32 {
+        a * 42 + b
+    }
+
+    #[export]
+    fn three(&mut self, _owner: Reference, a: i32, b: i32, c: i32) -> i32 {
+        a * 42 + b * c
+    }
+}
+
+fn test_variant_call_args() -> bool {
+    println!(" -- test_variant_call_args");
+
+    let ok = std::panic::catch_unwind(|| {
+        let obj = Instance::<VariantCallArgs>::new();
+
+        let mut base = obj.into_base().to_variant();
+
+        assert_eq!(
+            Some(42),
+            base.call(&"zero".into(), &[]).unwrap().try_to_i64()
+        );
+
+        assert_eq!(
+            Some(126),
+            base.call(&"one".into(), &[Variant::from_i64(3),])
+                .unwrap()
+                .try_to_i64()
+        );
+
+        assert_eq!(
+            Some(-10),
+            base.call(
+                &"two".into(),
+                &[Variant::from_i64(-1), Variant::from_i64(32),]
+            )
+            .unwrap()
+            .try_to_i64()
+        );
+
+        assert_eq!(
+            Some(-52),
+            base.call(
+                &"three".into(),
+                &[
+                    Variant::from_i64(-2),
+                    Variant::from_i64(4),
+                    Variant::from_i64(8),
+                ]
+            )
+            .unwrap()
+            .try_to_i64()
+        );
+    })
+    .is_ok();
+
+    if !ok {
+        godot_error!("   !! Test test_variant_call_args failed");
+    }
+
+    ok
+}
+
 fn init(handle: init::InitHandle) {
     handle.add_class::<Foo>();
     handle.add_class::<Bar>();
     handle.add_class::<RegisterSignal>();
+    handle.add_class::<VariantCallArgs>();
 }
 
 godot_gdnative_init!();

